### PR TITLE
GitHub#486 - Fix connection pool leaks

### PR DESCRIPTION
**Pull request by OwenLindsell**

### CODE REVIEW

File: components/client/src/main/java/com/hotels/styx/client/connectionpool/SimpleConnectionPool.java [hunk 1]
```diff
@@ -77,10 +77,12 @@ public Origin getOrigin() {
             Connection connection = dequeue();
             if (connection != null) {
                 borrowedCount.incrementAndGet();
+                sink.onCancel(() -> availableConnections.add(connection));
```
Comment (dvlato) [inline components/client/src/main/java/com/hotels/styx/client/connectionpool/SimpleConnectionPool.java]: Shouldn't we also reduce the borrowedCount as well?

File: components/client/src/main/java/com/hotels/styx/client/connectionpool/SimpleConnectionPool.java [hunk 2]
```diff
@@ -77,10 +77,12 @@ public Origin getOrigin() {
...
                 sink.success(connection);
             } else {
                 if (waitingSubscribers.size() < poolSettings.maxPendingConnectionsPerHost()) {
-                    this.waitingSubscribers.add(sink.onDispose(() -> waitingSubscribers.remove(sink)));
+                    this.waitingSubscribers.add(sink);
+                    sink.onDispose(() -> waitingSubscribers.remove(sink));
```
Comment (dvlato) [inline components/client/src/main/java/com/hotels/styx/client/connectionpool/SimpleConnectionPool.java]: I think this change can be reverted. However, I thought we had an issue with the waitingSubscribers list and we haven't changed anything in that code path... Don't we need to change anything here?

File: components/client/src/main/java/com/hotels/styx/client/connectionpool/SimpleConnectionPool.java [hunk 3]
```diff
@@ -143,6 +145,7 @@ private void queueNewConnection(Connection connection) {
             availableConnections.add(connection);
         } else {
             borrowedCount.incrementAndGet();
+            subscriber.onCancel(() -> availableConnections.add(connection));
```
Comment (dvlato) [inline components/client/src/main/java/com/hotels/styx/client/connectionpool/SimpleConnectionPool.java]: Same comment here about borrowedCount.

File: components/client/src/main/java/com/hotels/styx/client/connectionpool/SimpleConnectionPool.java (hunk 1 above)
Comment (OwenLindsell) [reply components/client/src/main/java/com/hotels/styx/client/connectionpool/SimpleConnectionPool.java]: Yes, very true.  I will fix!

Comment (OwenLindsell) [reply components/client/src/main/java/com/hotels/styx/client/connectionpool/SimpleConnectionPool.java]: Fixed

File: components/client/src/main/java/com/hotels/styx/client/connectionpool/SimpleConnectionPool.java (hunk 3 above)
Comment (OwenLindsell) [reply components/client/src/main/java/com/hotels/styx/client/connectionpool/SimpleConnectionPool.java]: Fixed

**Pull request merged**